docs/design/proposal/aws-sdk-upgrades.md

```diff
@@ -0,0 +1,142 @@
+  role for cross-account resource management
+* ACK code-generator depends on `aws-sdk-go` for generating K8s resource by
+  reading the AWS API models
+* ACK service controllers depend on `aws-sdk-go` to perform CRUD operations
+  using AWS API calls
+  
+## Background
+Before proposing new automation, it is important to understand how `aws-sdk-go`
+upgrades are handled in ACK repositories currently.
+#### Current steps
+* At first ACK `runtime` gets updated with the newer version of `aws-sdk-go` and
+  a new patch release is made for ACK `runtime`.
+* Then ACK `code-generator` is updated with new ACK runtime release. Because of
+  transitive dependency, the minimum required version of `aws-sdk-go` in 
+  code-generator gets updated to `aws-sdk-go` version in latest ACK runtime
+  release.
+* Then a new code-generator release is made, which triggers ACK service 
+  controller auto generation.
+* Controller auto-generation script updates the ACK `runtime` version in each 
+  controller to new ACK `runtime` version.
+* The update to new ACK `runtime` in service controller repositories updates the
+  minimum `aws-sdk-go` version in service controller `go.mod` file.
+* During auto-generation, ACK `code-generator` uses the `aws-sdk-go` present in
+  the service controller `go.mod` file to generate new service controller.
+
+The current process is automated for most parts but it misses following important
+parts:
+
+1. No upgrade trigger based on new `aws-sdk-go` releases
+2. No preview into the controller build. Currently ACK engineers have to manually
+   build cherry-picked controllers locally to see if new `aws-sdk-go` version
+   causes any regression. This is neither exhaustive nor automated.
+
+## High Level Overview
+
+The automation will be divided in following sub-parts
+
+### Trigger
+Trigger is responsible for when to upgrade ACK core libraries and service controllers
+with the new `aws-sdk-go` version.
+
+The trigger can be `event-driven` i.e. tracking a new release of `aws-sdk-go` using
+webhooks or other methods. 
+
+The trigger can also be `periodic`, where a periodic job constantly checks for the most
+recent release of `aws-sdk-go`, compares it with `aws-sdk-go` version in the ACK `runtime`.
+If most recent release is not same as ACK `runtime` dependency, then the automation
+gets triggered.
+
+### Verification
+This is where the automation verifies the new release and provide a preview of what
+the new changes will be for each repository. Example: Does a controller build successfully
+with the new `aws-sdk-go` version? Does a new resource/ new fields gets added as part
+of controller build? etc...
+
+### Release
+The actual process of upgrading the `go.mod` files with latest `aws-sdk-go` version
+and raising PRs.
+
+## Implementation Details
+
+### Trigger
+For the trigger, the periodic job option should be preferred that will
+run every day. 24 hours frequency is good enough to not delay verifying
+the new `aws-sdk-go` release. 
```

Comment:
Is this too frequently? I see from the [aws-sdk-go releases](https://github.com/aws/aws-sdk-go/releases) that we would be getting a new version most days. This would mean a new PR almost every day. Most new releases probably don't contain relevant changes to runtime or code-generator directly, and therefore we probably don't care to update runtime that often.

Is it possible to know whether a new version of the SDK impacts CARM or IAM?

Reply:
Excellent point!
I do agree that we should not update runtime on every patch release of `aws-sdk-go` (which seems almost everyday. I missed this).

Probably updating on every minor releases will be a good start (for runtime -> code-gen -> controllers) because I can see that for core SDK changes like session, endpoint etc, `aws-sdk-go` team cut a new minor release.

I will think more on handling specific patch release(Service Client Updates) for specific controllers.

Reply:
I'm hesitant to update the service controller's automatically. Or maybe teams could opt-in for that function. The reason is that a lot of service controller's will require updates to their hooks to support new spec fields properly. Probably most controller's shouldn't, for most of their resources, but some controllers (*cough S3 cough*) are notorious for specific fields having "custom logic". I'd love to get service team's feedback on how useful this would be for them.

Reply:
> I'm hesitant to update the service controller's automatically

Are you suggesting to only update core libraries?
If we update the ACK `runtime` with some `aws-sdk-go` version, service controllers will need to be updated with at least that `aws-sdk-go` version when they upgrade to latest runtime.

> a lot of service controller's will require updates to their hooks to support new spec fields properly

This is exact reason for adding verification step, which provides insight into the concerns you raised. And the automation just raises the PR, if existing functionality gets broken, i.e. e2e tests fail, the PR does not get merged. If any new resources get added, the PR does not get merged.